src/pmdas/amdgpu/amdgpu.c

```diff
@@ -0,0 +1,668 @@
+  uint32_t gpu_clock;
+  uint32_t mem_clock;
+  uint32_t temperature;
+  uint32_t load;
+  uint32_t avg_pwr;
+  int fd;
+  amdgpu_device_handle amd_device;
+} amdgpu_info_t;
+
+/* overall struct, holds instance values, indom and instance struct arrays */
+typedef struct {
+  uint32_t numcards;
+  uint32_t maxcards;
+  drmDevicePtr *devs;
+  amdgpu_info_t *info;
+  pmdaIndom *indom;
+} pcp_amdgpuinfo_t;
+
+static pcp_amdgpuinfo_t pcp_amdgpuinfo;
+static char mypath[MAXPATHLEN];
+static int isDSO = 1;
+static int drm_initialized;
+static int autorefresh = -1;
+static struct timeval interval;
+
+static int setup_gcard_indom(void) {
+  unsigned int device_count = 0;
+  pmdaIndom *idp = &indomtab[GCARD_INDOM];
+  char gpuname[32], *name;
+  int i, sts;
+
+  /* Initialize instance domain and instances. */
+  if ((sts = DRMDeviceGetDevices(&pcp_amdgpuinfo.devs,
+				      &pcp_amdgpuinfo.maxcards,
+				      &pcp_amdgpuinfo.numcards)) != DRM_SUCCESS) {
+    pmNotifyErr(LOG_ERR, "DrmDeviceGetDevies: %s", DRMErrStr(sts));
+    return sts;
+  }
+
+  device_count = pcp_amdgpuinfo.numcards;
+  pmNotifyErr(LOG_WARNING, "setup_gcard_indom: got %d cards", device_count);
+
+  pcp_amdgpuinfo.indom = idp;
+  pcp_amdgpuinfo.indom->it_numinst = 0;
+  pcp_amdgpuinfo.indom->it_set =
+      (pmdaInstid *)calloc(device_count, sizeof(pmdaInstid));
+
+  if (!pcp_amdgpuinfo.indom->it_set) {
+    pmNoMem("gcard indom", device_count * sizeof(pmdaInstid), PM_RECOV_ERR);
+    free(pcp_amdgpuinfo.devs);
+    return -ENOMEM;
+  }
+
+  if ((pcp_amdgpuinfo.info = (amdgpu_info_t *)calloc(
+           device_count, sizeof(amdgpu_info_t))) == NULL) {
+    pmNoMem("gcard values", device_count * sizeof(amdgpu_info_t), PM_RECOV_ERR);
+    free(pcp_amdgpuinfo.devs);
+    free(pcp_amdgpuinfo.indom->it_set);
+    return -ENOMEM;
+  }
+
+  for (i = 0; i < device_count; i++) {
+    drmDevicePtr dev = pcp_amdgpuinfo.devs[i];
+    amdgpu_info_t *info = &pcp_amdgpuinfo.info[i];
+
+    pcp_amdgpuinfo.indom->it_set[i].i_inst = i;
+    pmsprintf(gpuname, sizeof(gpuname), "gpu%d", i);
+    if ((name = strdup(gpuname)) == NULL) {
+      pmNoMem("gcard instname", strlen(gpuname), PM_RECOV_ERR);
+      while (--i)
+        free(pcp_amdgpuinfo.indom->it_set[i].i_name);
+      free(pcp_amdgpuinfo.devs);
+      free(pcp_amdgpuinfo.indom->it_set);
+      free(pcp_amdgpuinfo.info);
+      return -ENOMEM;
+    }
+    pcp_amdgpuinfo.indom->it_set[i].i_name = name;
+
+    info->fd = -1;
+    getAMDDevice(dev, &info->amd_device, &info->fd);
+    /* Get static values */
+    if (DRMDeviceGetGPUInfo(info->amd_device, &info->gpu_info)) {
+	info->failed[MEMORY_CLUSTER][AMDGPU_MEMORY_CLOCK_MAX] = 1;
+	info->failed[GPU_CLUSTER][AMDGPU_GPU_CLOCK_MAX] = 1;
+    }
+  }
+
+  pcp_amdgpuinfo.indom->it_numinst = device_count;
+  return 0;
+}
+
+static int refresh(pcp_amdgpuinfo_t *amdgpuinfo, uint32_t to_refresh)
+{
+  int i;
+
+  if (!drm_initialized) {
+      setup_gcard_indom();
+      drm_initialized = 1;
+  }
```

Comment:
I think we should only set drm_initialized here if setup_gcard_indom does not fail?  If it fails, we can try it again next time right?  (maybe an issue has resolved in the meantime)